### PR TITLE
PyTorch L1 regularization experiments

### PR DESCRIPTION
**Pull request by jjc2718**
After implementing logistic regression for TCGA pan-cancer gene alteration detection in PyTorch, I was confused about why L1 regularization wasn't giving me sparse regression coefficients. 

Turns out it is, but because I'm optimizing the model using stochastic gradient descent it's difficult to keep the coefficients at exactly 0 (see the Jupyter notebook for a more detailed explanation/investigation of why this is).

I'm a bit surprised that the 0 coefficients aren't similar between runs, though. I'm planning to run some simulations next to investigate this. Tentative plan is to take shuffled gene expression data and induce some correlation between features, but I need to think about the experimental design a bit.

### CODE REVIEW

File: pytorch_analysis.ipynb [hunk 1]
```diff
@@ -393,9 +404,318 @@
...
+    "        sk_genes = sk_coefs_folds['signal'][seed][fold][1]\n",
+    "        sk_nz_coefs = (sk_coefs != 0)\n",
+    "        sk_nz_genes = sk_genes[sk_nz_coefs]\n",
+    "        torch_coefs = torch_coefs_folds['signal'][seed][fold][0]\n",
+    "        torch_genes = torch_coefs_folds['signal'][seed][fold][1]\n",
+    "        torch_nz_coefs = (torch_coefs != 0)\n",
+    "        torch_nz_genes = torch_genes[torch_nz_coefs]\n",
+    "        inter_method_sims.append(compute_jaccard(set(sk_nz_genes), set(torch_nz_genes)))\n",
+    "        \n",
+    "import itertools as it\n",
```
Comment (arielah) [inline pytorch_analysis.ipynb]: This block of code seems very similar (but not identical) to the same block 50 lines down. Could they be combined into a single loop? If nothing else, you shouldn't have to import itertools twice.

File: pytorch_analysis.ipynb [hunk 2]
```diff
@@ -393,9 +404,318 @@
...
+    "\n",
+    "Jaccard similarity\n",
+    "\n",
+    "$$J(A, B) = \\frac{|A \\cap B|}{|A \\cup B|}$$\n",
+    "\n",
+    "A and B here are the sets of genes with nonzero coefficients, so this just looks at binary overlap since the scale of the coefficients doesn't matter.\n",
+    "\n",
+    "### **Takeaways:**\n",
+    "\n",
+    "Coefficients are not very stable at all between methods. Despite having similar predictive performance and similar sparsity, the nonzero coefficients found by PyTorch and sklearn have almost no overlap.\n",
```
Comment (arielah) [inline pytorch_analysis.ipynb]: Nice walk through of the results throughout the notebook and a good summary at the end.

File: pytorch_vs_sklearn.py [hunk 3]
```diff
@@ -40,8 +41,30 @@
                help='where to write results to')
 p.add_argument('--seed', type=int, default=cfg.default_seed)
 p.add_argument('--verbose', action='store_true')
+
+prms = p.add_argument_group('pytorch_params')
```
Comment (arielah) [inline pytorch_vs_sklearn.py]: I've never seen an argument group used in argparse. I'm curious more than anything, why create a group instead of just passing the params individually? It seems like you call them the same way (e.g. args.batch_size)?

Comment (jjc2718) [reply pytorch_vs_sklearn.py]: I mostly used it because it groups the arguments together in the help output (e.g. when I call `python pytorch_vs_sklearn.py -h`), making it obvious that they have a related function.

There might be other things you can do with them, I'm not sure.

File: pytorch_analysis.ipynb (hunk 1 above)
Comment (jjc2718) [reply pytorch_analysis.ipynb]: I moved this logic into a utilities file and got rid of the duplication - let me know if that's what you had in mind.
